theano/tensor/alt_gemm_template.c

```diff
@@ -0,0 +1,133 @@
+/** %(name)s **/
+void alt_numpy_scalar_matrix_product_in_place_%(float_type)s(%(float_type)s scalar, PyArrayObject* matrix) {
+    NpyIter* iterator = NpyIter_New(matrix, 
+        NPY_ITER_READWRITE | NPY_ITER_EXTERNAL_LOOP | NPY_ITER_REFS_OK, 
+        NPY_KEEPORDER, NPY_NO_CASTING, NULL);
+    if(iterator == NULL)
+        alt_fatal_error("Unable to iterate over a matrix "
+                        "for a scalar * matrix operation.");
+    NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterator, NULL);
+    char** data_ptr = NpyIter_GetDataPtrArray(iterator);
+    npy_intp* stride_ptr = NpyIter_GetInnerStrideArray(iterator);
+    npy_intp* innersize_ptr = NpyIter_GetInnerLoopSizePtr(iterator);
+    do {
+        char* data = *data_ptr;
+        npy_intp stride = *stride_ptr;
+        npy_intp count = *innersize_ptr;
+        while(count) {
+            *((%(float_type)s*)data) *= scalar;
+            data += stride;
+            --count;
+        }
+    } while(get_next(iterator));
+    NpyIter_Deallocate(iterator);
+}
+/*Matrix+Matrix function. Compute (coeffA * matrixA) + (coeffB * matrixB)
+ * Remark: This function actually sums a C-contiguous matrix (alpha*op(A)*op(B)) with a F-contiguous matrix (beta*C)
```

Comment:
I think I was confused by the fact that you were mentioning "C", which is not a parameter of this function, but of `gemm`. I think we should not refer to the parameters of gemm, but only of this function in the comment, and reference the actual name of the parameters (for instance, "ALPHA" and "A" and not "coeffA" and "matrixA").
We should also say that it actually does that computation inplace in B, something like `B = ALPHA * A + BETA * B`.

Regarding the memory layout of the matrices, does this function actually need A to be C-contiguous and B to be F-contiguous, or even C order vs F order? Or should it work either way? Because if it is a limitation of this function, we should add some checks at the beginning of the function. If not, then we can reformulate the comment to say that it would be a performance limitation if called with matrices with different layouts.

Also, if you do the same trick with `PyArray_MatrixProduct` that you did with `PyArray_MatrixProduct2` (changing the transposed status of A and B and swapping them), then the transpose of the resulting array would be in F order, like C.